### ISSUE TITLE
statement construction for multi-scope rules

### ISSUE THREAD

**Issue by yelhamer** [mandiant/capa#1643]
This issue is in order to discuss how statements should be constructed for rules that have 2 scopes, static and dynamic. Should we enforce rules to be composed of features common to both scopes? should we create two separate statements and use each one where required? and most importantly, how should we treat legacy rules?

### Option 1:
The first option is to enforce rule authors to use only common features in such mixed rules, which would result in the same statement for both scopes, and therefore rules would have just one statement to pass around.

**cons:**
- restrictive for rule authors: no mixed features (static and dynamic) in each rule.
- we would need to either go back and modify old rules (at their features' level), or make them available only for static use by setting the dynamic scope to `InvalidScope` and create dynamic equivalents of them.

**pros:**
- simple to implement

### Option 2:
The second option is to create two separate statements, one for the static scope, and one for the dynamic one, with each statement being made up only of its valid features. Then, we would use each statement when necessary.

**cons:**

- it at least doubles the number of statements in the ruleset, which might make things slower (thoughts?)
- we must ensure that we're never passing the wrong statement around, since that might create some bugs that might be hard to detect (rules not matching because the statement is incompatible).

**pros:**
- in this scenario, rules should have mixed features (static and dynamic). This means that when constructing each statement we would prune features that are of the other flavor, which means that we should be able to support static rules to some extent in the dynamic flavor.

### Option 3:
The third option would be to merge the two statements from the previous approach into one large inclusive statement. This statement would include features that valid for either of the rule's scopes, and would be passed in whole to the matching engine. When the `match()` function is doing the matching, non-valid features would just be ignored, since they couldn't have been extracted anyways, and non-valid scopes (static ones in a dynamic analysis run) would not match.

**cons:**
- it feels like a bit of a hack

**pros:**
- easier to implement and maintain: no pruning involved, and no multi-statements either.
- it would be easy to support old rules.
- if we choose to make it possible to specify different scopes in the future at subscope level, then this approach should make things easier. For example:
```yaml
rule:
  meta:
    scopes:
      static: function
      dynamic: thread
  features:
    - and:
      - instruction|call:
        - and:
          - api: some_api
          - or:
            - arguments:
              - description: api callscope arguments
              - arg1: 0x123
              - arg2: 0x456
            - and:
              - description: api arguments
              - number: 0x123
              - number: 0x456
```

### Supporting old rules:
Currently, we're enforcing all instances of the `Rule` class to have both scopes specified (child rules have the `DEV_SCOPE` sinkhole scope), and I think it's imperative to make it possible for rules to be able to specify just one scope (at least for subscope rules).

I propose that we make the `scopes` keyword the standard, and provide a script (or a sed oneliner really) that replaces "scope: {some scope}" with "scopes:\n  static: {some scope}". However, supporting the legacy format of "scope:" also sounds fine to me.

I think that we should also assume the dynamic scope to be `process` when no dynamic scope is specified; and if the rule should be only used in static analysis, then the rule author should specify the dynamic scope as `invalid` as @williballenthin suggested previously.

thoughts on all of this?

**Comment by mr-tz** [comment]
We should try to keep it simple and 3 sounds good to me for that. I'm not a fan of the `instruction|call` definition though I'd have to see more rules in practice.

**Comment by williballenthin** [comment]
i agree that (3) makes sense, in fact, it may require the fewest code changes. let's try this out, and if we encounter unexpected complications, we can re-evaluate and use (2) as a fallback.

**Comment by williballenthin** [comment]
> I propose that we make the scopes keyword the standard, and provide a script (or a sed oneliner really) that replaces "scope: {some scope}" with "scopes:\n static: {some scope}". However, supporting the legacy format of "scope:" also sounds fine to me.

if we're going to update rules, then we should do a little extra work to make the scopes explicit. that means providing both `static:` and `dynamic:` entries. only if we're going to continue to support `scope:` should we have implicit behavior.

updating all rules with sed seems easy, though it doesn't really take into account finding rules that don't make sense in a dynamic context. do we make the changes blindly and then fix things up by hand? or make the changes slowly, only as we've had a chance to review the logic? im inclined to recommend the second one until we're sure what we want to do.

if we used sed to make the changes but not commit them, then we could use a tool like `tig status` (or other interactive git UI) to review the rules and the scope mapping and easily stage/unstage/revert them. it would still require going through 800+ rules.

so again im feeling like we should maybe migrate a subset of rules that are interesting and relevant and see how that goes. only once we're happy will i be willing to spend hours reviewing rule updates.

**Comment by yelhamer** [comment]
+1 for using only a subset of rules to test things out. i suggest we create a dynamic-feature-extraction branch on capa-rules and then add the rules we think would be nice to test. once we're satisfied with things, we can merge that branch to master.

**Comment by yelhamer** [comment]
Closed in #1671
